test/test_cuda_sanitizer.py

```diff
@@ -334,6 +334,39 @@ def test_multiple_wait(self):
         self.assert_good_kernel_launch(stream_id(2), read_only=[tensor_id(1)])
         self.assert_good_kernel_launch(stream_id(3), read_only=[tensor_id(1)])
 
+    def test_device_synchronize(self):
+        # Tests that a device synchronization does correctly cause all streams
+        # to synchronize with each other.
+
+        iterations = 10
+        for i in range(1, iterations):
+            self.assert_good_kernel_launch(stream_id(i), read_write=[tensor_id(i)])
+
+        self.handler._handle_device_synchronization()
+        self.assert_good_kernel_launch(
+            stream_id(0), read_write=[tensor_id(i) for i in range(1, iterations)]
+        )
+
+    def test_stream_synchronize(self):
+        self.assert_good_kernel_launch(stream_id(1), read_write=[tensor_id(1)])
+        self.handler._handle_stream_synchronization(stream_id(1))
+
+        self.assert_good_kernel_launch(stream_id(2), read_only=[tensor_id(1)])
+        self.assert_good_kernel_launch(stream_id(3), read_only=[tensor_id(1)])
```

Comment:
do you need any bad_kernel_launch tests here?

Reply:
I've added a `bad_kernel_launch` to highlight the distinction between stream and device synchronization, and added some new tests as well.